ext/flash/lib.rs

```diff
@@ -0,0 +1,1451 @@
+  let _ = sock.write(&response);
+
+  #[cfg(unix)]
+  {
+    use std::os::unix::io::AsRawFd;
+    if let InnerStream::Tcp(stream_handle) = &sock.inner {
+      let stream_handle = stream_handle.as_raw_fd();
+      if let Some(fd) = resource.clone().backing_fd() {
+        // SAFETY: all-zero byte-pattern is a valid value for libc::stat.
+        let mut stat: libc::stat = unsafe { std::mem::zeroed() };
+        // SAFETY: call to libc::fstat.
+        if unsafe { libc::fstat(fd, &mut stat) } >= 0 {
+          let _ = sock.write(
+            format!("Content-Length: {}\r\n\r\n", stat.st_size).as_bytes(),
+          );
+          let tx = sendfile::SendFile {
+            io: (fd, stream_handle),
+            written: 0,
+          };
+          tx.await?;
+          return Ok(());
+        }
+      }
+    }
+  }
+
+  let _ = sock.write(b"Transfer-Encoding: chunked\r\n\r\n");
+  loop {
+    let vec = vec![0u8; 64 * 1024]; // 64KB
+    let buf = ZeroCopyBuf::new_temp(vec);
+    let (nread, buf) = resource.clone().read_return(buf).await?;
+    if nread == 0 {
+      let _ = sock.write(b"0\r\n\r\n");
+      break;
+    }
+    let response = &buf[..nread];
+
+    let _ = sock.write(format!("{:x}", response.len()).as_bytes());
+    let _ = sock.write(b"\r\n");
+    let _ = sock.write(response);
+    let _ = sock.write(b"\r\n");
+  }
+
+  resource.close();
+  Ok(())
+}
+
+pub struct RespondChunkedFast;
+
+impl fast_api::FastFunction for RespondChunkedFast {
+  fn function(&self) -> *const c_void {
+    op_flash_respond_chunked_fast as *const c_void
+  }
+
+  fn args(&self) -> &'static [fast_api::Type] {
+    &[
+      fast_api::Type::V8Value,
+      fast_api::Type::Uint32,
+      fast_api::Type::TypedArray(fast_api::CType::Uint8),
+      fast_api::Type::Bool,
+    ]
+  }
+
+  fn return_type(&self) -> fast_api::CType {
+    fast_api::CType::Void
+  }
+}
+
+unsafe fn op_flash_respond_chunked_fast(
+  recv: v8::Local<v8::Object>,
+  token: u32,
+  response: *const fast_api::FastApiTypedArray<u8>,
+  shutdown: bool,
+) {
+  let ptr =
+    recv.get_aligned_pointer_from_internal_field(V8_WRAPPER_OBJECT_INDEX);
+  let ctx = &mut *(ptr as *mut ServerContext);
+
+  let response = &*response;
+  if let Some(response) = response.get_storage_if_aligned() {
+    respond_chunked(ctx, token, shutdown, Some(response));
+  } else {
+    todo!();
+  }
+}
+
+fn respond_chunked(
+  ctx: &mut ServerContext,
+  token: u32,
+  shutdown: bool,
+  response: Option<&[u8]>,
+) {
+  let sock = match shutdown {
+    true => {
+      let tx = ctx.response.remove(&token).unwrap();
+      tx.socket()
+    }
+    // In case of a websocket upgrade or streaming response.
+    false => {
+      let tx = ctx.response.get(&token).unwrap();
+      tx.socket()
+    }
+  };
+
+  if let Some(response) = response {
+    let _ = sock.write(format!("{:x}", response.len()).as_bytes());
+    let _ = sock.write(b"\r\n");
+    let _ = sock.write(response);
+    let _ = sock.write(b"\r\n");
```

Comment:
multiple syscalls?

Reply:
Ideally we’d use write_all_vectored() here.
Also I think we shouldn’t ignore errors.